### ISSUE TITLE
Provide option for FTRL Optimizer to not reset variable value as zero when its first gradient is zero

### ISSUE THREAD

**Issue by wwiiiii** [tensorflow/tensorflow#50118]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.5.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Under current implementation of FTRL optimizer, the value of the variable becomes zero regardless of its original value when its gradient is zero on initial `apply_gradients`.

Reproducible snippet:

```python
opt = tf.keras.optimizers.Ftrl(learning_rate=0.1, initial_accumulator_value=0.1)
x = tf.Variable([-2, -1, 0, 1, 2], dtype=tf.float32)

with tf.GradientTape() as tape:
    loss = tf.math.reduce_sum(tf.nn.relu(x))

grads = tape.gradient(loss, [x])


print('Before applying gradients', x.numpy())
print('Gradients:', grads)
opt.apply_gradients(zip(grads, [x]))
print('After applying gradients', x.numpy())
```

Result:
```
Before applying gradients [-2. -1.  0.  1.  2.]
Gradients: [<tf.Tensor: shape=(5,), dtype=float32, numpy=array([0., 0., 0., 1., 1.], dtype=float32)>]
After applying gradients [0.        0.        0.        0.6031424 1.301631 ]
```
The first three values became zero even their gradients were zero.

This behavior is not a bug as we initialize all model parameters as zero in the original paper. This was not a problem for linear models, but is problematic for deep models due to the increased importance of initial weight distribution.

We can support its usage for deep models as well by providing an option to prevent this phenomenon.

**Will this change the current api? How?**
Yes, we may add a new boolean parameter in the constructor of `tf.keras.optimizers.Ftrl`.

**Who will benefit with this feature?**
Whoever uses FTRL optimizer for their deep models.

**Any Other info.**
I guess we can fix it by initializing the value of the `linear` slot based on the `variable` value when the slot is first created.
If it's desired change, I can prepare PR to fix it.

**Comment by fchollet** [comment]
Thanks for reaching out.

> Who will benefit with this feature?
> Whoever uses FTRL optimizer for their deep models.

The FTRL optimizer was designed for shallow models (logistic regression over very large feature sets, in the context of ad targeting). I would recommend *not* using it for a deep model.

Because of this, we wouldn't be inclined to add an option to implement the behavior you describe.

As a workaround, are you able to make your own subclass of FTRL to implement the behavior you need?

**Comment by wwiiiii** [comment]
Although the FTRL optimizer was originally designed for linear models, it can also be used with deep models after making modifications as mentioned above. It is mathematically equivalent to the Adagrad optimizer in this case if other regularization coefficients(L1/L2 ) are all zero.

Especially when using this, we can expect more practical (compared to L1 loss) sparsity constraint on network parameters as it sets the value of the parameter satisfying the regularization condition to exactly 0 during training. I believe other built-in optimizers don't have such pruning behavior.

Anyway, if not many people will benefit from this, maybe it would be better to write custom code rather than modifying FTRL class in keras. Please close this issue if that's the case.

**Comment by ymodak** [comment]
Closing this issue for now. Feel free to reopen if further discussion/information is available. Thanks!
